torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -195,6 +197,18 @@ std::unique_ptr<RpcCommandBase> RequestCallbackImpl::processRpc(
 
       return c10::guts::make_unique<PropagateGradientsResp>();
     }
+    case MessageType::CLEANUP_AUTOGRAD_CONTEXT_REQ: {
+      auto& cleanupContextReq = static_cast<CleanupAutogradContextReq&>(rpc);
+      auto cleanupContextId = cleanupContextReq.getContextId();
+      // release the context if it still exists on this thread. We need to check
+      // if it exists since it may have been deleted by an in-flight RPC.
+      if (DistAutogradContainer::getInstance().hasContextWithId(
+              cleanupContextId)) {
+        DistAutogradContainer::getInstance().releaseContext(
```

Comment:
Actually, we probably need to do this check + release atomically, since it is still possible that another thread could release the context after we've done the `hasContextWithId` check. Then, `releaseContext` would throw but we don't necessarily want to throw, since the context has been freed as desired. Should we make a version of releaseContext that does not throw if the context we are releasing does not exist (or control this with a param?)

Reply:
Yes, let's do check + release automatically and have a version which doesn't throw if the context is cleaned up (can call it releaseContextIfPresent).

Reply:
Also, shouldn't `notifyWorkers` be true here? In case we do nested RPCs, we need this node to notify other nodes too.

Reply:
It should eventually be true, but since this PR doesn't implement cycle detection this is a way of not doing additional RPC calls. Once #27576 lands we can modify this to support nested rpc calls.

Reply:
Do we need to do cycle detection? If we do an atomic check + release, we might ask a node to release again, but that would just be a noop right?

Reply:
Yes, this is true. Though for performance reasons we'd probably want to avoid the extra RPC caused by asking a node to release again.

Reply:
> Yes, this is true. Though for performance reasons we'd probably want to avoid the extra RPC caused by asking a node to release again.

True, although for now let's just go with a simple version and later on we can evaluate whether we need to optimize this further.